src/iterative_ensemble_smoother/sies.py

```diff
@@ -0,0 +1,416 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Optional, Union
+
+import numpy as np
+import scipy as sp  # type: ignore
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
+from iterative_ensemble_smoother.sies_inversion import (
+    inversion_direct,
+    inversion_subspace_exact,
+    inversion_subspace_projected,
+)
+from iterative_ensemble_smoother.utils import _validate_inputs, sample_mvnormal
+
+
+class SIES:
+    """
+    Initialize a Subspace Iterative Ensemble Smoother (SIES) instance.
+
+    This is an implementation of the algorithm described in the paper:
+    Efficient Implementation of an Iterative Ensemble Smoother for Data Assimilation and Reservoir History Matching
+    written by Evensen et al (2019),
+    URL: https://www.frontiersin.org/articles/10.3389/fams.2019.00047/full
+
+    Parameters
+    ----------
+    parameters : npt.NDArray[np.double]
+        A 2D array of shape (num_parameters, ensemble_size). Each row corresponds
+        to a parameter in the model, and each column corresponds to an ensemble
+        member (realization). This is X in Evensen (2019).
+    covariance : npt.NDArray[np.double]
+        Either a 1D array of diagonal covariances, or a 2D covariance matrix.
+        The shape is either (num_parameters,) or (num_parameters, num_parameters).
```

Comment:
The shape here should be (num_observations,) or (num_observations, num_observations).